ripd/ripd.c

```diff
@@ -3056,15 +3041,159 @@ DEFUN (show_ip_rip,
-				} else
+				} else {
 					vty_out(vty, "self            ");
+				}
+
 				vty_out(vty, "%3" ROUTE_TAG_PRI,
 					(route_tag_t)rinfo->tag);
 			}
 
 			vty_out(vty, "\n");
 		}
 	}
+}
+
+static void rip_show_vty_json(struct vty *vty, struct rip *rip,
+			      json_object *json)
+{
+	struct route_node *np;
+	struct rip_info *rinfo = NULL;
+	struct list *list = NULL;
+	struct listnode *listnode = NULL;
+	json_object *json_routes = NULL;
+
+	json_object_int_add(json, "vrfId", rip->vrf->vrf_id);
+	json_object_string_add(json, "vrfName", rip->vrf->name);
+
+	if (!json_routes) {
+		json_routes = json_object_new_object();
+		json_object_object_add(json, "routes", json_routes);
+	}
+
+	for (np = route_top(rip->table); np; np = route_next(np)) {
+		list = np->info;
+
+		if (!list)
+			continue;
+
+		json_object *json_route = NULL;
+		json_object *json_path = NULL;
+
+		for (ALL_LIST_ELEMENTS_RO(list, listnode, rinfo)) {
+			if (!json_route) {
+				json_route = json_object_new_array();
+				json_object_object_addf(json_routes, json_route,
+							"%pFX", &np->p);
+			}
+
+			json_path = json_object_new_object();
+			json_object_array_add(json_route, json_path);
+
+			switch (rinfo->nh.type) {
+			case NEXTHOP_TYPE_IPV4:
+			case NEXTHOP_TYPE_IPV4_IFINDEX:
+				json_object_string_addf(json_path, "nexthop",
+							"%pI4",
+							&rinfo->nh.gate.ipv4);
+				json_object_int_add(json_path, "metric",
+						    rinfo->metric);
+				break;
+			case NEXTHOP_TYPE_IFINDEX:
+				json_object_string_add(json_path, "nexthop",
+						       "0.0.0.0");
+				json_object_int_add(json_path, "metric",
+						    rinfo->metric);
+				break;
+			case NEXTHOP_TYPE_BLACKHOLE:
+				json_object_string_add(json_path, "nexthop",
+						       "blackhole");
+				json_object_int_add(json_path, "metric",
+						    rinfo->metric);
+				break;
+			case NEXTHOP_TYPE_IPV6:
+			case NEXTHOP_TYPE_IPV6_IFINDEX:
+				json_object_string_add(json_path, "nexthop",
+						       "V6 Address Hidden");
+				json_object_int_add(json_path, "metric",
+						    rinfo->metric);
+				break;
+			}
+
+			/* Route which exist in kernel routing table. */
+			if ((rinfo->type == ZEBRA_ROUTE_RIP) &&
+			    (rinfo->sub_type == RIP_ROUTE_RTE)) {
+				json_object_string_addf(json_path, "from",
+							"%pI4", &rinfo->from);
+				json_object_int_add(json_path, "tag",
+						    rinfo->tag);
+				rip_vty_out_uptime(vty, rinfo, json_path);
+			} else if (rinfo->metric == RIP_METRIC_INFINITY) {
+				json_object_string_add(json_path, "from",
+						       "self");
+				json_object_int_add(json_path, "tag",
+						    rinfo->tag);
+				rip_vty_out_uptime(vty, rinfo, json_path);
+			} else {
+				if (rinfo->external_metric)
+					json_object_string_add(json_path,
+							       "from", "self");
+				else
+					json_object_string_add(json_path,
+							       "from", "self");
+
+				json_object_int_add(json_path, "tag",
+						    rinfo->tag);
+			}
+		}
+	}
+}
+
+DEFUN (show_ip_rip,
```

Comment:
Nit: we could use `DEFPY` instead of `DEFUN`, but I don't think it is a blocker.

Reply:
I left this function as it was, the diff shows (looks like) that I created it, but it's just moved.